File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.py
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        
        def inorder_traversal(root, remaining):
            if not root:
                return None

            left_result = inorder_traversal(root.left, remaining)
            if left_result is not None:
                return left_result
            remaining[0] -= 1
            if remaining[0] == 0:
                return root.val

            return inorder_traversal(root.right, remaining)
        remaining = [k]
        return inorder_traversal(root, remaining)

File: 0230-kth-smallest-element-in-a-bst/0230-kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public ArrayList<Integer> inorder(TreeNode root, ArrayList<Integer> res) {
        if (root == null) return res;
        inorder(root.left, res);
        res.add(root.val);
        inorder(root.right, res);
        return res;
    }

    public int kthSmallest(TreeNode root, int k) {
        ArrayList<Integer> res = inorder(root, new ArrayList<Integer>());
        return res.get(k-1);
    }
}